src/common/array_sort.fypp

```diff
@@ -21,19 +21,21 @@ SUBROUTINE ${prefix}$_sort(arr, n, indices)
    integer, DIMENSION(1:n), INTENT(INOUT)   :: indices
 
    INTEGER                                  :: i
-   ${type}$, POINTER                        :: tmp_arr(:)
-   INTEGER, POINTER                         :: tmp_idx(:)
+   ${type}$, ALLOCATABLE                    :: tmp_arr(:)
+   INTEGER, ALLOCATABLE                     :: tmp_idx(:)
 
-   IF (n == 0) RETURN ! for some reason this is a frequent case in cp2k
+   IF (1 < n) THEN
+      ! scratch space used during the merge step
+      ALLOCATE(tmp_arr((n+1)/2), tmp_idx((n+1)/2))
 
-   ! scratch space used during the merge step
-   ALLOCATE(tmp_arr((size(arr)+1)/2), tmp_idx((size(arr)+1)/2))
+      indices = (/ (i, i = 1, n) /)
 
-   indices = (/ (i, i = 1, size(arr)) /)
+      CALL ${prefix}$_sort_low(arr(1:n), indices, tmp_arr, tmp_idx)
 
-   CALL ${prefix}$_sort_low(arr(1:n), indices, tmp_arr, tmp_idx)
-
-   DEALLOCATE(tmp_arr, tmp_idx)
+      DEALLOCATE(tmp_arr,tmp_idx)
+   ELSE IF (0 < n) THEN ! can be a frequent case in cp2k
```

Comment:
I'd prefer: 
```
IF (n==1) THEN
  indices(1) = 1
ELSE IF (n>1) THEN
   ...
END IF
```

Reply:
Sure, can do. It was actually ordered by "likelyhood" (but mispredicting the branch is the least problem). Sorting appears to be not a hotspots (otherwise we should probably revisit allocating the scratch array recursively and rather only allocate the whole index set upfront and sort without temporary memory).

Although there is a SORT intrinsic in Fortran, it appears to be unable to indirectly sort. A quick C-call would probably cut-it (even C-sort allows taking the comparator as a function pointer and hence it can be used to sort indirectly without any temp storage other than the upfront index array).

If you want, I can go for the latter...

Reply:
Thanks, but I think our current solution is sufficient. I'm also not aware that sort showed up as hotspot.

I was just concerned with readability. If one reads `ELSE IF (0<n) indices(1) = 1` without the context then it looks rather confusing.

Reply:
The `IF (0 < n)` is some "good style" adopted from C/C++. The "rule" is to have the constant value in front. The reasoning is `if (a == 1)` is sometimes (incorrectly/typo) written as `if (a = 1)`, which compiles in an unwanted fashion.

Reply:
https://en.wikipedia.org/wiki/Yoda_conditions

Reply:
@dev-zero Cool, I did not know this!